lib/basedriver/commands/execute.js

```diff
@@ -0,0 +1,139 @@
+import _ from 'lodash';
+import log from '../logger';
+import vm from 'vm';
+import { attach } from 'webdriverio';
+import { MJSONWP_ELEMENT_KEY, W3C_ELEMENT_KEY } from '../../protocol/protocol';
+
+const BASE_PATH = '/wd/hub'; // TODO defining this here is brittle but it is hardcoded in routes, so...
+const FEAT_FLAG = 'execute_driver_script';
+const SCRIPT_TYPE_WDIO = 'webdriverio';
+// TODO add wd script type
+
+let commands = {};
+
+/**
+ * This method takes a string which is executed as javascript in the context of
+ * a new nodejs VM, and which has available a webdriverio driver object, having
+ * already been attached to the currently running session.
+ *
+ * @param {string} script - the string representing the driver script to run
+ * @param {string} [scriptType='webdriverio'] - the name of the driver script
+ * library (currently only webdriverio is supported)
+ *
+ * @returns {Object} - a JSONifiable object representing the return value of
+ * the script
+ */
+commands.executeDriverScript = async function (script, scriptType = 'webdriverio') {
+  if (!this.isFeatureEnabled(FEAT_FLAG)) {
+    throw new Error(`Execute driver script functionality is not available ` +
+                    `unless server is started with --allow-insecure including ` +
+                    `the '${FEAT_FLAG}' flag, e.g., --allow-insecure=${FEAT_FLAG}`);
+  }
+
+  if (scriptType !== SCRIPT_TYPE_WDIO) {
+    throw new Error(`Only the '${SCRIPT_TYPE_WDIO}' script type is currently supported`);
+  }
+
+  if (!this.opts.host || !this.opts.port) {
+    throw new Error('Host or port of running server were not defined; this ' +
+                    'is required. This is probably a programming error in the driver');
+  }
+
+  const logLevels = ['error', 'warn', 'log'];
+  const logs = {};
+  const consoleFns = {};
+  for (const level of logLevels) {
+    logs[level] = [];
+    consoleFns[level] = (...logMsgs) => logs[level].push(...logMsgs);
+  }
+
+  const driver = attach({
+    sessionId: this.sessionId,
+    protocol: 'http', // Appium won't ever be behind ssl locally
+    hostname: this.opts.host,
+    port: this.opts.port,
+    path: BASE_PATH,
+    isW3C: this.isW3CProtocol(),
+    isMobile: true,
+    capabilities: this.caps
+  });
+  const fullScript = buildScript(script);
+  let result = await vm.runInNewContext(fullScript)(driver, consoleFns);
```

Comment:
so, do we want to limit the maximum script execution time?

Reply:
I think it's a good idea, though an infinite loop will not be prevented by doing so. I guess this should be modulated by a cap?

Reply:
This can be also an argument to the endpoint. But it also makes sense to have a default value of, let say 60 minutes.

Reply:
oh yeah, endpoint argument makes more sense.

Reply:
ok i spent a bunch of time trying to implement this, and it doesn't work, because the client script is executed in a promise, so it avoids the node vm timeout. i tried to implement timeouts using promises, but Promise.race causes the whole process to hang presumably because one of the promises is never read. i'm not sure. I think we have 2 options:

1) forget any useful timeout argument and just make clear to the sys admin that a single user could cause an appium server to stall using this

2) run all this in a subproc, not a nodejs vm, so we can kill it from the outside if it's being bad.

what do you think?

Reply:
i decided to work on a combination of proc fork + vm, we'll see...

Reply:
ok, i got it working. this is a bit slower since it has to spin up a new node process for each request, but it will be vastly safer i guess!

Reply:
(it will even work if someone implements an infinite loop)